### DISCUSSION TITLE
Item inventory

### DISCUSSION THREAD

**Discussion by Naeta2** [peter-kish/gloot#136] [Q&A]
Is there a way to create an item that have itself its own inventory ? I'm working on a system where the player has no "global inventory", and their inventory slots are tied to the clothes / bags they have equipped. Is this currently possible with gloot ? If not it's a suggestion then (: 
thanks for your work on the addon, it is a great timesaver !

**Comment by peter-kish** [comment]
In the current version - no, not directly. But you can probably find a workaround, something similar to what you described in [this discussion](https://github.com/peter-kish/gloot/discussions/127).

With that said, I'm currently working on item slots that can "own" the items (https://github.com/peter-kish/gloot/issues/130). That way the items would be stored inside the slot directly and there would be no need for an inventory at all.

**Reply by Naeta2** [reply]
great news, thank you very much ! (:

**Comment by Naeta2** [comment]
Quick follow up on me trying to modify gloot to implement something like this.
I managed to link items to inventory and it looks somewhat like this for now :

<details>
  <summary>Image</summary>
  
![image](https://github.com/peter-kish/gloot/assets/46900232/3430e7b3-e9a8-4fc7-b3d2-b0aee95a19a8)

</details>

I can move items in and out of an item's inventory, close the inventory, change an item's slot, it will all be saved.

If someone, one day, wants to do the same thing and/or improve it (I am not very good at godoting), here is what I did : 

First I made a InventoriesData singleton, which only contains an initially empty dictionnary variable called GlobalInventories. There is also a PlayerData singleton with a player_inventory dictionnary. This is specific to my game and this dictionnary contains itself dictionnaries where the data of the multiple inventory slots on the left are stored. Not ideal, I know, but couldn't think of a better way

I then edited the default inventory_item.gd in gloot's files (I wanted to create a custom item class but I did not manage to make the protoset use another item class, so I edited the original script) to add these lines : 

<details>
  <summary>Code</summary>
  
  ```
var global_inventories = InventoriesData.GlobalInventories

@export var inventory : InventoryGrid

#container_item properties :
#is_container -> bool
#container_width -> int
#container_height -> int
#inventory_id -> int

func _ready() :
	global_inventories = InventoriesData.GlobalInventories
	if get_property("is_container", false) :
		connect_inventory()
		if get_property("inventory_id", null) == null :
			set_global_id()
		else :
			retrieve_inventory_data()

func connect_inventory() :
	var new_inv = InventoryGrid.new()
	add_child(new_inv)
	inventory = new_inv
	inventory.item_protoset = protoset
	inventory.size = Vector2(get_property("container_width", 1), get_property("container_height", 1))

func retrieve_inventory_data() :
	if inventory.deserialize(global_inventories[get_property("inventory_id", null)]) != null :
		inventory.deserialize(global_inventories[get_property("inventory_id", null)])

func close_inventory():
	if global_inventories.has(get_property("inventory_id")) :
		global_inventories[get_property("inventory_id", null)] = inventory.serialize()

func set_global_id():
	var i = global_inventories.size()
	set_property("inventory_id", i)
	global_inventories[global_inventories.size()] = inventory.serialize()
  ```
  
</details>


as you see, I use a boolean item property "is_container". It is not required on usual items but you must add it to your items that should have an inventory. You must also add a "container_width" int and a "container_height" int.

Then it is just a matter of detecting if an item is a container, if yes retrieve its inventory data from the GlobalInventories singleton using its "inventory_id" property, and instancing a CtrlInventory UI linked to the inventory automatically created by the item when it is instanced

I also added a line in inventory.gd under can_add_item so that you cannot store an object in itself : 

<details>
  <summary>Code</summary>

```
if item == get.parent():
	return false
```

</details>

I noticed that with that method it creates lots of empty dictionnaries in GlobalInventories. I searched why and from what I understand GLoot seems to be creating some sort of "proxy" item before replacing it by the "real" item with its saved properties when I deserialize my dictionnary, and so as this proxy item does not have a property "inventory_id" yet, it creates a new empty one, and when the proxy is replaced by the real item, we can interact with the items saved and all seem normal, but the dictionnary keeps these empty inventories that point to nowhere and it keeps adding them up. Not sure if this is due to my code or to GLoot's code ?

I also noticed that when I am transfering an item from an item with an inventory to my equipment slots, the item preview disappears when I'm dragging it, but the item appears in the destination inventory when I drop it, same thing I'm not sure why it does it

ps : Peter would you have a discord account or somewhere I could send you messages without bothering you ? Sometimes I'm like "hey I'd like to ask how this thing works or how it does this" but often it's too small or specific to make a post / comment about it here

**Reply by peter-kish** [reply]
Ok, so this is quite a complex description of the issue(s) and I'm not sure I understand what exactly is going on :smiley: 
so I would suggest coming up with a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example#:~:text=In%20computing%2C%20a%20minimal%20reproducible,to%20be%20demonstrated%20and%20reproduced.) which only makes minimal calls to the addon and demonstrates the issue.
Otherwise I'd have to familiarize myself with other people's code for each issue before I can figure out how they are using the addon and help them out :face_with_spiral_eyes: 

I have a discord account (kish7126) but I don't use it too often and might be more responsive via email (pkish023@gmail.com).

**Reply by Naeta2** [reply]
I was putting a little project to demonstrate the issue and I found it myself, it was on my end : I was putting an item in the inventory via the godot editor, so each time the inventory was opened in the game, it first loaded that default item I put in the editor and so following my code it gave it a new id before deserializing the dictionary and putting the "real" item in the inventory. 
Second issue seem to be on my end too, I looked into it yesterday and I think it's about the ui layers, my guess is the dragged item must be in a layer behind the other half of my ui
So it's all good I think, my items with inventories seem to work, no issue. I'll keep this habit of trying to replicate the issue with minimal code, I still have lots of things to learn (: 
Thanks !
